src/mug/tls.gleam

```diff
@@ -0,0 +1,630 @@
+///
+/// For maximum interoperability, the certificates in the chain should be in the correct order, as the chain will be 
+/// sent as-is to the peer. If chain certificates are not provided, certificates from the configured trusted CA certificates 
+/// will be used to construct the chain.
+///
+/// If `NoVerification` is set, this function does nothing.
+///
+pub fn certs_keys(
+  options: TlsConnectionOptions,
+  certs_keys certs_keys: List(CertsKeys),
+) {
+  TlsConnectionOptions(
+    ..options,
+    certificates: case options.certificates {
+      Certificates(system, cacerts: cacerts, ..) ->
+        Certificates(
+          use_system_cacerts: system,
+          cacerts: cacerts,
+          certs_keys: certs_keys,
+        )
+      NoVerification -> NoVerification
+    },
+  )
+}
+
+/// Establish a TLS-encrypted TCP connection to the server specified in the
+/// connection options.
+///
+/// Returns an error if the connection could not be established.
+///
+/// The socket is created in passive mode, meaning the the `receive` function is
+/// to be called to receive packets from the client. The
+/// `receive_next_packet_as_message` function can be used to switch the socket
+/// to active mode and receive the next packet as an Erlang message.
+///
+pub fn connect(options: TlsConnectionOptions) -> Result(Socket, Error) {
+  let host = charlist.from_string(options.host)
+  ssl_connect(
+    host,
+    options.port,
+    get_tls_options(options.certificates),
+    options.timeout,
+  )
+}
+
+/// Upgrade a plain TCP connection to TLS.
+///
+/// Accepts a TCP socket and performs the client-side TLS handshake. This
+/// may not work on all TCP servers.
+///
+/// Returns an error if the connection could not be established.
+///
+/// The socket is created in passive mode, meaning the the `receive` function is
+/// to be called to receive packets from the client. The
+/// `receive_next_packet_as_message` function can be used to switch the socket
+/// to active mode and receive the next packet as an Erlang message.
+///
+/// The `certs_keys` option can be 
+///
+pub fn upgrade(
+  socket: TcpSocket,
+  certificates certificates: Certificates,
+  milliseconds timeout: Int,
+) -> Result(Socket, Error) {
+  ssl_upgrade(socket, get_tls_options(certificates), timeout)
+}
+
+/// Whether to use system certificates or not.
+pub fn use_system_cacerts(system: Bool) {
+  Certificates(system, None, [])
+}
+
+fn get_tls_options(
+  certificates: Certificates,
+) -> List(#(TlsOptionName, Dynamic)) {
+  [
+    // When data is received on the socket queue it in the TCP stack rather than
+    // sending it as an Erlang message to the socket owner's inbox.
+    #(Active, dynamic.from(False)),
+    // We want the data from the socket as bit arrays please, not lists.
+    #(Mode, dynamic.from(Binary)),
+    #(
+      Verify,
+      dynamic.from(case certificates {
+        NoVerification -> VerifyNone
+        _ -> VerifyPeer
+      }),
+    ),
+  ]
+  |> fn(opts) {
```

Comment:
Never pipe into a function please

Reply:
I do not know why I did it like that haha. Fixed.